network-config-analyzer/NetworkConfigQuery.py

```diff
@@ -614,12 +615,174 @@ def exec(self):
 
         return QueryAnswer(True, f'{self.name1} and {self.name2} are semantically equivalent.')
 
+class SemanticDiffQuery(TwoNetworkConfigsQuery):
+    """
+    Produces a report of newly allowed and newly denied connections
+    """
+    class SingleDiff:
+        """
+        Representing a single diff between a pair of eps
+        """
+        def __init__(self, from_ep, to_ep, removed, added):
+            self.from_ep = from_ep
+            self.to_ep = to_ep
+            self.removed = removed
+            self.added = added
+
+        def __str__(self):
+            res = "\n" + str(self.from_ep) + " -> " + str(self.to_ep) + ":\n"
+            res += "\tRemoved: " + str(self.removed) + "\n"
+            res += "\tAdded: " + str(self.added)
+            return res
+
+    @staticmethod
+    def pretty_print_diff(diff_list):
+        """
+        pretty printing the results of semantic diff
+        :param diff_list:
+        :return:
+        """
+        result = ""
+        for single in diff_list:
+            if diff_list[single]:
+                result += "- " + single + str(diff_list[single]) + "\n"
+                result = result.replace('{', ' ').replace('}', ' ').replace('\'', '')
```

Comment:
This is ugly. The three calls to `replace` scan the whole result string every iteration. They should only work on the added string.

Reply:
good catch. done